samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -117,8 +117,9 @@
    * Samza SystemProducer implementation.
    *
    * @param sinkFn the function to send messages in this stream to an external system
+   * @return the input {@link MessageStream}
    */
-  void sink(SinkFunction<? super M> sinkFn);
+  MessageStream<M> sink(SinkFunction<? super M> sinkFn);
```

Comment:
Can we add more documentation around what message is propagated to the downstream operators and its characteristics within the scope of the `SinkFunction`? 

- E.g. It is immutable within the scope of the sink function. 
- When is the message propagated to the downstream? Is it after the completion of SinkFunction? What if the sink function is asynchronous like remote call or kafka send? Do we guarantee any sort of happens before?